proof/refine/ARM/Ipc_R.thy

```diff
@@ -5223,26 +5274,268 @@ lemma replyPush_sch_act_wf:
   by (wpsimp wp: sts_sch_act' hoare_vcg_all_lift hoare_vcg_if_lift hoare_drop_imps)
 
 lemma replyPush_invs':
-  "\<lbrace>\<lambda>s. invs' s \<and> sch_act_not callerPtr s \<and> st_tcb_at' active' callerPtr s \<and>
-        ex_nonz_cap_to' callerPtr s \<and> ex_nonz_cap_to' calleePtr s \<and> ex_nonz_cap_to' replyPtr s
-        \<and> tcb_at' calleePtr s \<and> reply_at' replyPtr s
-        \<and> obj_at' (\<lambda>ko. replyTCB ko = None) replyPtr s\<rbrace>
+  "\<lbrace>\<lambda>s. invs' s \<and> sch_act_not callerPtr s \<and> obj_at' (\<lambda>ko. replyTCB ko = None) replyPtr s \<and>
+        tcb_at' calleePtr s \<and> st_tcb_at' (\<lambda>st. st \<noteq> Inactive \<and> st \<noteq> IdleThreadState) callerPtr s \<and>
+        ex_nonz_cap_to' callerPtr s \<and> ex_nonz_cap_to' calleePtr s \<and> ex_nonz_cap_to' replyPtr s\<rbrace>
    replyPush callerPtr calleePtr replyPtr canDonate
    \<lbrace>\<lambda>_ s. invs' s\<rbrace>"
   unfolding invs'_def valid_state'_def valid_pspace'_def
   apply (wpsimp wp: replyPush_valid_objs' replyPush_sch_act_wf replyPush_valid_queues
                     replyPush_if_live_then_nonz_cap' replyPush_valid_idle'
               simp: valid_pspace'_def)
-  apply (subgoal_tac "invs' s", fastforce)
-  by (simp add: invs'_def valid_state'_def valid_pspace'_def)
+  apply (fastforce simp: invs'_def valid_state'_def valid_pspace'_def
+                         pred_tcb_at'_def obj_at'_def projectKOs
+                  intro: invs_valid_replies')
+  done
 
-lemma setThreadState_Running_invs':
-  "\<lbrace>\<lambda>s. invs' s \<and> tcb_at' t s \<and> ex_nonz_cap_to' t s\<rbrace>
-   setThreadState Running t
+lemma setEndpoint_invs':
+  "\<lbrace>invs' and valid_ep' ep and ex_nonz_cap_to' eptr\<rbrace> setEndpoint eptr ep \<lbrace>\<lambda>_. invs'\<rbrace>"
+  by (wpsimp simp: invs'_def valid_state'_def comp_def)
+
+lemma setSchedContext_valid_ep'[wp]:
+  "setSchedContext param_a param_b \<lbrace>valid_ep' ep\<rbrace>"
+  unfolding valid_ep'_def
+  by (cases ep; wpsimp wp: hoare_vcg_ball_lift)
```

Comment:
Something like `set_sc’.typ_at_sc_at’_n_lifts’` can probably work instead of this, but I am not too sure.